File: Design-Patterns-Using-Kotlin/State/StatePattern2.kt
interface State {
    fun handle(context: Context)
}

class Context(private var currentState: State) {
    fun changeState(state: State) {
        this.currentState = state
    }

    fun request() {
        currentState.handle(this)
    }
}

class ConcreteState1 : State {
    override fun handle(context: Context) {
        println("ConcreteState1 handle")
        context.changeState(ConcreteState2())
    }
}

class ConcreteState2 : State {
    override fun handle(context: Context) {
        println("ConcreteState2 handle")
        context.changeState(ConcreteState1())
    }
}

// Client code.
fun main() {
    val state1: State = ConcreteState1()
    val context = Context(state1)
    context.request()
    context.request()
}

/*
ConcreteState1 handle
ConcreteState2 handle
*/

File: Design-Patterns-Using-Cpp/State/StatePattern2.cpp
#include <iostream>

class Context;

class State {
public:
    virtual void handle(Context& context) = 0;
    virtual ~State() {}
};

class Context {
private:
    State* currentState;

public:
    Context(State* state) : currentState(state) {}

    void changeState(State* state) {
        currentState = state;
    }

    void request() {
        currentState->handle(*this);
    }
};


class ConcreteState1 : public State {
public:
    void handle(Context& context) override;
};

class ConcreteState2 : public State {
public:
    void handle(Context& context) override;
};


void ConcreteState1::handle(Context& context) {
    std::cout << "ConcreteState1 handle" << std::endl;
    context.changeState(new ConcreteState2());
}

void ConcreteState2::handle(Context& context) {
    std::cout << "ConcreteState2 handle" << std::endl;
    context.changeState(new ConcreteState1());
}

// Client code
int main() {
    State* state1 = new ConcreteState1();
    Context context(state1);

    context.request();
    context.request();

    delete state1;  // Freeing allocated memory

    return 0;
}

/*
ConcreteState1 handle
ConcreteState2 handle
*/